Mana preview

rudehero opened the pull request:
Created a small arrow which will point along the mana bar to indicate the amount of mana a skill will use.  This is accomplished via a series of signals between the circular menu and the manabar.

There are many more updated files than I'd expected, I believe due to godot's processing/compilation processes.  If there is a way to avoid such noisy pull requests to simplify the reviewer's job please let me know.

Appreciate any/all feedback.

NathanLovato commented:
You shouldn't remove the *.import files, they're important. Then, when you stage files in git, you shouldn't do `git add *` unless you checked with `git status` that you do want them all.
You'll have to go edit your commit, with e.g. `git rebase -i HEAD^` (rebase interactively up until one commit before the current one the head is pointing to, generally the last commit). This allows you to edit commits, remove staged files.

NathanLovato commented on godot/combat/interface/circular_menu/CircularMenu.gd:
```diff
@@ -21,13 +25,22 @@ func initialize(actor : Battler) -> void:
 		button.initialize(action, target_position)
 		button.connect("pressed", self, "_on_CircularButton_pressed", [action])
+		if action is SkillAction:
+			button.connect("focused", self, "_on_CircularButton_focused", [action.skill.mana_cost])
```
You could emit signals directly instead of adding two callback methods: `button.connect("focused", self, "emit_signal", ["button_focused", [action.skill.mana_cost]])` (please check that passing arguments this way works, I'm not sure)

rudehero replied:
This is a much more elegant solution and worked great.

NathanLovato commented on godot/combat/interface/circular_menu/CircularMenu.gd:
```diff
@@ -21,13 +25,22 @@ func initialize(actor : Battler) -> void:
+	for bar in bars:
+		if bar is ManaBar:
+			manabar = bar
+			manabar._connect_preview(self)
```
You're calling a private method on a separate object. You shouldn't ever do that

rudehero replied:
I think this was an issue of changing my intention.  I don't believe that should have been a private function and have renamed it to reflect.  

However, if you disagree I can try to find a different way to accomplish this.  I saw mention in the guidelines of a global events handler, but it doesn't appear to be implemented at this time.

NathanLovato commented on godot/combat/interface/circular_menu/CircularMenu.gd:
```diff
@@ -89,9 +102,19 @@ func _unhandled_input(event: InputEvent) -> void:
 func _on_CircularButton_pressed(action):
+	emit_signal("disable_preview_mana")
 	yield(close(), "completed")
+	manabar._disconnect_preview(self)
```
Calling private method

NathanLovato commented on godot/combat/interface/circular_menu/CircularButton.gd:
```diff
@@ -39,8 +45,14 @@ func enter_focus():
 	tooltip.hide()
+	emit_signal("unfocused")
 	animation_player.play('deactivate')
+	yield(animation_player,"animation_finished")
```
Does any other have object wait for this method to return? If not, you shouldn't yield

rudehero replied:
Correct.  Remnant from debugging something.  Removed.

rudehero replied:
Actually, did some double checking and that line of code wasn't mine.  You can find it in the current master branch.  Removing it caused animation issues when cycling through buttons.

NathanLovato replied:
It's strange that it appears as a new line in the PR. But got it, thanks

NathanLovato commented on godot/combat/interface/bars/manabar/ManaBar.gd:
```diff
@@ -3,10 +3,35 @@ the inherited variables from the base class """
 extends "res://combat/interface/bars/StatBar.gd"
 class_name ManaBar
```
If nothing else uses the class, you shouldn't register it.

rudehero replied:
Within circular menu there is a check being conducted against the class to ensure only the manabar is getting connected.  However, I'm changing this behavior to more general.  I'm modifying the base StatBar to accept connections, this will allow for alternate resource bar types to include HP as a resource.  This additionally removes the need for registering the class.

NathanLovato replied:
Thanks!

NathanLovato commented on godot/combat/interface/bars/manabar/ManaBar.gd:
```diff
@@ -3,10 +3,35 @@ the inherited variables from the base class """
 	self.value = battler_stats.mana
+
+func _connect_preview(menu : Control) -> void:
+	menu.connect("preview_mana", self, "_on_preview_mana")
```
Please define the signals at the top of the script

rudehero replied:
Refactored, connected signals were added to the top of StatBar.gd

NathanLovato commented on godot/combat/interface/bars/manabar/ManaBar.gd:
```diff
@@ -3,10 +3,35 @@ the inherited variables from the base class """
+	if cost > self.value:
+		return
+
+	var mana_size = ($Column/TextureProgress.rect_size.x - 4) / self.max_value
```
The nodes this script depends upon should be stored in `onready var` at the top of the script. Also, if you can make calculations in advance, please do so.

rudehero replied:
Added the onready var.
Adjusted to have as much math done once within initialization as possible.

NathanLovato commented:
If you're having trouble with the git rebase, we can have a call on discord and I can guide you.

rudehero commented on godot/combat/interface/circular_menu/CircularMenu.gd:
```diff
@@ -21,13 +25,22 @@ func initialize(actor : Battler) -> void:
 	"""
 	Creates a circular menu from a battler's actions
 	"""
+	active_actor = actor
```
Correct.  Removed.

rudehero replied:
This variable was in use, but was redundant and has been removed.

rudehero commented on godot/combat/interface/circular_menu/CircularMenu.gd:
```diff
@@ -1,9 +1,13 @@
 extends Control
 
 signal action_selected(action)
+signal preview_mana(mana)
```
Renamed two signals:
button_focused
button_unfocused

rudehero commented:
Thank you for the feedback, apologies that first attempt was rather messy with so many files.  Rebased and cleaned up, hopefully easier to work through now.  

Made changes based on feedback as well as topics discussed in the guidelines linked.

NathanLovato commented:
Thanks for the improvements. You don't have to comment on every change, github will mark review points as outdated once you change the corresponding lines.

NathanLovato commented on godot/combat/interface/bars/StatBar.gd:
```diff
@@ -29,12 +33,27 @@ func set_value(new_value) -> void:
 func _on_value_changed(new_value, old_value) -> void:
 	self.value = new_value
 	
+func _on_button_focused(cost) -> void:
```
Why are there these two empty methods?

NathanLovato commented:
Open RPG is getting a Godot 4 rewrite, closing because of this. See https://github.com/GDQuest/godot-open-rpg/pull/209

Thanks for your work!
